### DISCUSSION TITLE
Make AlsCharacter inherit from a custom Character class

### DISCUSSION THREAD

**Discussion by oserrat** [Sixze/ALS-Refactored#173] [Q&A]
Right now I have my PlayerCharacter inherit from AlsCharacter, but I want to create a common Character class, so ALS inherits from this for my player and also I'm able to create other characters without ALS. 

![image](https://user-images.githubusercontent.com/56392636/199350052-bd13c3f3-c00e-4c19-b87a-7456f2e18ced.png)

Changing ACharacter with my custom class gives me a circular dependency error between ALS.Build.cs and MyProject.Build.cs

Is there any way to move AAlsCharacter class from the plugin to my project?

**Comment by ameaninglessname** [comment]
I think you can put your "common stuff" into a  plugin, or turning these functionality into components

**Reply by oserrat** [reply]
Thank you, I've put my core classes into a separate CoreModule and made ALS inherit from that, then the player class in my GameModule inherits from ALS. The only thing that I'm having trouble with is that I'm not able to use ActorComponents (defined in the CoreModule) in the GameModule. Do you have any idea why?

**Reply by ameaninglessname** [reply]
What's the error message?
Did you reference the module in your game module's build.cs?

**Reply by oserrat** [reply]
yes it's in the build.cs, its not a compiling error. The ActorComponent object pointer is null, so I cannot access to that component even though It is being created correctly in the class constructor. It's like the object reference is getting lost after the constructor because the Component is being created on GameModule instead of CoreModule

This is the declaration
![image](https://user-images.githubusercontent.com/56392636/200052530-90b51837-654c-4c7f-b85a-56fe1a20910d.png)

This is at runtime in the class constructor
![image](https://user-images.githubusercontent.com/56392636/200053312-9cee516e-bb06-4f84-87ff-2b28a6bb3b57.png)

**Reply by ameaninglessname** [reply]
> yes it's in the build.cs, its not a compiling error. The ActorComponent object pointer is null, so I cannot access to that component even though It is being created correctly in the class constructor. It's like the object reference is getting lost after the constructor because the Component is being created on GameModule instead of CoreModule
> 
> This is the declaration ![image](https://user-images.githubusercontent.com/56392636/200052530-90b51837-654c-4c7f-b85a-56fe1a20910d.png)
> 
> This is at runtime in the class constructor ![image](https://user-images.githubusercontent.com/56392636/200053312-9cee516e-bb06-4f84-87ff-2b28a6bb3b57.png)

This should not happen. Did you call DestoryComponent on Inventory somewhere?
You can put a break point in it, after the constructor.
Also you can check if inventory exist when begin play.

**Reply by oserrat** [reply]
I found this two posts related to this issue:
https://forums.unrealengine.com/t/inherited-component-details-not-showing/388642/7
https://forums.unrealengine.com/t/inherited-component-details-not-showing-in-blueprint/42671/59?page=3

Reparenting class to Actor and then to its original class and fixed the problem. Also just changing the name of the component so Unreal forces to recompile every reference.
